### PR TITLE
Mesh sensor value print assert

### PR DESCRIPTION
**Pull request by m-alperen-sener**

### CODE REVIEW

**Review by de-nordic — APPROVED** [review]

File: subsys/bluetooth/mesh/sensor.c [hunk 1]
```diff
@@ -623,7 +623,7 @@ const char *bt_mesh_sensor_ch_str(const struct sensor_value *ch)
 {
 	static char str[BT_MESH_SENSOR_CH_STR_LEN];
 
-	bt_mesh_sensor_ch_to_str(ch, str, sizeof(str));
+	__ASSERT_NO_MSG(bt_mesh_sensor_ch_to_str(ch, str, sizeof(str)) == 0);
```
Comment (alxelax) [inline subsys/bluetooth/mesh/sensor.c:626]: I'm not sure that API that parses customer data should fall into assertion. The overall devices become fragile due to customer incoming data.

Should it return a predefined string instead? Smth like:
```
const char *bt_mesh_sensor_ch_str(const struct sensor_value *ch)
{
	static char str[BT_MESH_SENSOR_CH_STR_LEN];

	if (bt_mesh_sensor_ch_to_str(ch, str, sizeof(str)) == 0) {
		strcpy(str, "Unknown format");
	}

	return str;
}
```

Comment (PavelVPV) [inline subsys/bluetooth/mesh/sensor.c:626]: Are you sure it is correct? Aren't `bt_mesh_sensor_ch_to_str` which calls `snprintk` returns the number of bytes written? Also, if asserts are disabled, `bt_mesh_sensor_ch_to_str` will be removed: https://github.com/zephyrproject-rtos/zephyr/blob/a04a1d8991e312ece8530f0ab35b444010113fb9/include/zephyr/sys/__assert.h#L134-L143. We should never put function calls into assert macros.

**Review by PavelVPV — CHANGES_REQUESTED** [review]

**Comment by PavelVPV** [comment]
Please also add to the commit message the exact issue that caused this change.

**Comment by de-nordic** [comment]
> Please also add to the commit message the exact issue that caused this change.

```
[2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:35: error: '%06u' directive output may be truncated writing between 6 and 10 bytes into a region of size between 7 and 17 [-Werror=format-truncation=]

[2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",

[2023-09-15T15:09:53.987Z]       |                                   ^~~~~~~~~~~

[2023-09-15T15:09:53.987Z] <redacted>zephyr/include/zephyr/sys/printk.h:67:32: note: in definition of macro 'snprintk'

[2023-09-15T15:09:53.987Z]    67 | #define snprintk(...) snprintf(__VA_ARGS__)

[2023-09-15T15:09:53.987Z]       |                                ^~~~~~~~~~~

[2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h: In function 'bt_mesh_sensor_ch_str':

[2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:41: note: format string is defined here

[2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",

[2023-09-15T15:09:53.987Z]       |                                         ^~~~

[2023-09-15T15:09:53.987Z] In function 'bt_mesh_sensor_ch_to_str',

[2023-09-15T15:09:53.987Z]     inlined from 'bt_mesh_sensor_ch_str' at <redacted>nrf/subsys/bluetooth/mesh/sensor.c:626:2:

[2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:35: note: directive argument in the range [0, 2147483647]

[2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",

[2023-09-15T15:09:53.987Z]       |                                   ^~~~~~~~~~~

[2023-09-15T15:09:53.987Z] <redacted>zephyr/include/zephyr/sys/printk.h:67:32: note: in definition of macro 'snprintk'

[2023-09-15T15:09:53.987Z]    67 | #define snprintk(...) snprintf(__VA_ARGS__)

[2023-09-15T15:09:53.987Z]       |                                ^~~~~~~~~~~

[2023-09-15T15:09:53.987Z] <redacted>zephyr/include/zephyr/sys/printk.h:67:23: note: 'snprintf' output between 9 and 23 bytes into a destination of size 19

[2023-09-15T15:09:53.987Z]    67 | #define snprintk(...) snprintf(__VA_ARGS__)

[2023-09-15T15:09:53.987Z]       |                       ^~~~~~~~~~~~~~~~~~~~~

[2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:16: note: in expansion of macro 'snprintk'

[2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",

[2023-09-15T15:09:53.987Z]       |                ^~~~~~~~

[2023-09-15T15:09:53.987Z] cc1: all warnings being treated as errors
```

**Comment by PavelVPV** [comment]
> > Please also add to the commit message the exact issue that caused this change.
> 
> ```
> [2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:35: error: '%06u' directive output may be truncated writing between 6 and 10 bytes into a region of size between 7 and 17 [-Werror=format-truncation=]
> 
> [2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",
> 
> [2023-09-15T15:09:53.987Z]       |                                   ^~~~~~~~~~~
> 
> [2023-09-15T15:09:53.987Z] <redacted>zephyr/include/zephyr/sys/printk.h:67:32: note: in definition of macro 'snprintk'
> 
> [2023-09-15T15:09:53.987Z]    67 | #define snprintk(...) snprintf(__VA_ARGS__)
> 
> [2023-09-15T15:09:53.987Z]       |                                ^~~~~~~~~~~
> 
> [2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h: In function 'bt_mesh_sensor_ch_str':
> 
> [2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:41: note: format string is defined here
> 
> [2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",
> 
> [2023-09-15T15:09:53.987Z]       |                                         ^~~~
> 
> [2023-09-15T15:09:53.987Z] In function 'bt_mesh_sensor_ch_to_str',
> 
> [2023-09-15T15:09:53.987Z]     inlined from 'bt_mesh_sensor_ch_str' at <redacted>nrf/subsys/bluetooth/mesh/sensor.c:626:2:
> 
> [2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:35: note: directive argument in the range [0, 2147483647]
> 
> [2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",
> 
> [2023-09-15T15:09:53.987Z]       |                                   ^~~~~~~~~~~
> 
> [2023-09-15T15:09:53.987Z] <redacted>zephyr/include/zephyr/sys/printk.h:67:32: note: in definition of macro 'snprintk'
> 
> [2023-09-15T15:09:53.987Z]    67 | #define snprintk(...) snprintf(__VA_ARGS__)
> 
> [2023-09-15T15:09:53.987Z]       |                                ^~~~~~~~~~~
> 
> [2023-09-15T15:09:53.987Z] <redacted>zephyr/include/zephyr/sys/printk.h:67:23: note: 'snprintf' output between 9 and 23 bytes into a destination of size 19
> 
> [2023-09-15T15:09:53.987Z]    67 | #define snprintk(...) snprintf(__VA_ARGS__)
> 
> [2023-09-15T15:09:53.987Z]       |                       ^~~~~~~~~~~~~~~~~~~~~
> 
> [2023-09-15T15:09:53.987Z] <redacted>nrf/include/bluetooth/mesh/sensor.h:484:16: note: in expansion of macro 'snprintk'
> 
> [2023-09-15T15:09:53.987Z]   484 |         return snprintk(str, len, "%s%u.%06u",
> 
> [2023-09-15T15:09:53.987Z]       |                ^~~~~~~~
> 
> [2023-09-15T15:09:53.987Z] cc1: all warnings being treated as errors
> ```

Looks like `BT_MESH_SENSOR_CH_STR_LEN` should just be set to 23 here: https://github.com/nrfconnect/sdk-nrf/blob/68117767e79e40a9018221a30daa0f27e053af0b/include/bluetooth/mesh/sensor.h#L33
@m-alperen-sener , could you try this?

**Comment by m-alperen-sener** [comment]
I have removed the assert with the constant message. By design the sensor value can not be bigger then 19 chars, sign + val1 + val2 <= 19.

**Comment by PavelVPV** [comment]
> I have removed the assert with the constant message. By design the sensor value can not be bigger then 19 chars, sign + val1 + val2 <= 19.
> 
> Any value greater than 0 should be fine for return.

Isn't it better to 4 more bytes to the array and always print the value rather than adding additional function call and some text?

**Comment by ludvigsj** [comment]
> Isn't it better to 4 more bytes to the array and always print the value rather than adding additional function call and some text?

The problem is that if someone has e.g. "20020030" as val2 in their sensor_value, it does not make sense to print it because we _know_ that something weird has happened to the sensor_value.

File: subsys/bluetooth/mesh/sensor.c (hunk 1 above)
Comment (m-alperen-sener) [reply subsys/bluetooth/mesh/sensor.c]: I have removed the assert for the case build with no asserts.

**Review by PavelVPV — APPROVED** [review]

**Review by alxelax — APPROVED** [review]

**Pull request merged**